README
MP3info	- Adds, deletes, or changes MP3 ID3 tags.


ABOUT
=====

Q: What's MP3Info?

A: MP3Info is a little utility used to read and modify the ID3 tags in MPEG 
layer 3 (MP3) files.  In case you don't know, MP3 is a very popular digital 
audio format which encodes a recording with a lossy compression algorithm 
resulting in files representing a minute of near-CD quality stereo sound in 
less than a megabyte of data.   An ID3 tag is a little scrap of data added to 
the end of the MP3 file which stores information about the audio recording 
including the title, artist, album, copyright date, and musical genre.  


HISTORY
=======

MP3Info was originally written by Ricardo Cerqueira <rmc@rccn.net>.  As of 
January 2000, MP3Info is being maintained by Cedric Tefft 
<cedric@earthling.net>.  If you have any bug reports or patches, please email 
them to me.  Bugs get fixed a lot faster if you include patches -- especially 
if the problem is platform specific.


PORTABILITY
===========

MP3Info was developed under Linux, but should work on most UN*X variants.
Included with the text-based version is a gui version which requires the GTK
libraries (see http://www.gtk.org).  The console version runs in plain
command-line and interactive curses mode.  The curses or ncurses library is
required.


INSTALLATION
============

Pretty basic.  After unpacking the archive run the following two commands:

make
make install 		(this command must be run by the root user)

If you want to compile just the console or GTK version use 'make mp3info' or 
'make gmp3info' (respectively).


USAGE: Console Version
======================

mp3info [-h | -G] (-i | -d | -t title -a artist -l album -y year -c comment
                             -g genre) MP3_FILE [ MP3_FILE [ ... ]]

-h      --      Display this info and exit
-G      --      Display genre list and exit
-i      --      Edit ID3 tag interactively
-d      --      Delete ID3 tag

  Specify ID3 tag fields:

-t  title      --     Track title
-a  artist     --     Artist Name
-l  album      --     Album Name
-y  year       --     Copyright Year
-c  comment    --     Comment
-g  genre      --     Genre (use -G option for a list of valid genres)

Non-specified fields, if existant, will remain unchanged.
Genres can be specified as numbers or names: '-g 17' same as '-g Rock'
Remember: Multiple word fields must be enclosed in "" (eg: -t "A title")

Usage: GTK Version
==================

gmp3info song.mp3


EXAMPLES
========

Display existing ID3 tag information (if any) in song.mp3

	mp3info song.mp3

Set the title, author and genre of song.mp3. (All other fields unchanged)

	mp3info -t "Song Title" -a Author -g "Rock & Roll" song.mp3

Set the album field of all MP3 files in the current directory 
to "The White Album"

	mp3info -a "The White Album" *.mp3

Delete the ID3 tag from song1.mp3 and song2.mp3

	mp3info -d song1.mp3 song2.mp3


KNOWN BUGS
==========

There's no "save and quit" in interactive mode. You must fill in all the 
fields (even if it is with blanks) and let the program finish by itself. 
CTRL+C does leave MP3info, but the data isn't saved.

Using space to erase tags in interactive mode does not work correctly if you 
then backspace over the deleted text.

Some valid MP3 files are rejected as being invalid. This is a thorny one.  If 
you figure it out, please let me know.


KNOWN LIMITATIONS
=================

The title, author, and album fields are limited to 30 characters.  This is a 
limitation of the ID3 tag format, not MP3Info.

Genres cannot be specified arbitrarily.  They must be specified from a 
pre-determined list (use mp3info -G to see that list).  Again, this is a 
limitation of the ID3 tag format.


MISCELLANEOUS NEWS & INFO
=========================

MP3Info is looking for a new name!  Several people have pointed out that there 
are other, older programs out there called mp3info which arguably have more of 
a "right" (in as much as such a thing exists in the open source world) to the 
name than this program.  Name suggestions, like bug reports, are happily 
accepted.


DISCLAIMER
==========

I do not take responsibility for damaged files, MP3s or not. If you decide to 
run this on something other than an MP3, tough luck. You're on your own. It 
shouldn't cause any damage, but if it does, blame yourself!


TO DO
=====

Fix the known bugs? :-)

Add an "Open file" option to GMP3info.



					- Cedric Tefft <cedric@earthling.net>


ChangeLog
v0.6.1  - Fixed a bug that set genre to 'Blues' if no genre was given on the command 
	  line and the MP3 file did not have an existing ID3 tag.

v0.6	- As of this version maintenance of mp3info taken over by Cedric Tefft <cedric@earthling.net>
	- Reworked command-line/curses version to handle processing multiple MP3's.
	- Added -d (delete ID3 tag) switch
	- Added full text/numeric genre handling to command-line, curses and GTK modes.
	- Added -G option (display genre table)
	- View mode is now determined automatically when no other options are specified.  
	  The -v option is accepted for backward compatibility, but ignored.  It 
	  should probably be removed in a future version.
	- Added ID3 structure to make internal argument passing a little cleaner.
	- Consolidated ID3 tag reading/writing code.
	- Interactive mode now displays existing tags.
	- Modified Makefile to make gmp3info dependant on gmp3info.c

v0.5	- Added genre tag. (Still in numerical form) (Thanks to jsc@arsdigita.com)
	- Corrected a byte ordering bug which affected Sun systems (Thanks to Benjamin Korvemaker)
	- Corrected a permissions bug (Thanks to Carsten Hahn)
	- Fixed a GTK bug in gmp3info. (Thanks go to Thomas Schultz)
v0.4	- gmp3info is now part of the pack.
	  Fixed some incompatibity problems with SGI (thanks again to Petter) and RedHat 6.0.
	- Added MP3 check, using the magic number (Thanks to Steven James for this one)

v0.3.1	- Fixed a couple of HP/UX incompatibilities (Thanks to Petter Reinholdtsen)
	- Added -v (to view an existing tag)

v0.3	- Fixed "trash in the fields" bug. (Strings are now forced to be null-terminated)
	  Updated help screen.
	  Fixed minor warnings during compile.
	  If "-h" is among options, others are now ignored.
	  "-i" now must be specified for interactive (curses) mode.
	  Commented and cleaned code.

v0.3-b1 - Added reading and preservation of existing tags in command line mode.
	  Fixed occasional tag seek bug.
	  Changed options parsing to use "getopt"

v0.3-a1 - Merged the command line version back in.
	  Fixed "tag appending instead of overwriting" bug

v0.2-b1 - Fixed "non-existant file is created" bug.
	  Added a help screen
	  Field lengths became constants.

v0.2-a2	- A few minor bugfixes.
	- Added ^C support (terminal was going crazy if the program was interrupted)
	- Added flexible window size (window width is always equal to terminal width)

v0.2-a1	- Not released
	- curses version. Weeeeeee! (Way better :-) )

v0.1.1  - Not released
	- Added support for the "Year" and "Comment" fields.
	- Fixed quite a few overflow problems

v0.1 	- Initial version. Never released


mp3curs.c
/*
 * mp3curs.c
 *
 * Curses functions for MP3info
 *
 * Original Author:
 *       Ricardo Cerqueira      <rmc@rccn.net>
 *
 * This software is covered by the GNU public license, which should be in a file named
 * LICENSE acompanying this.
 *
 */

#include "mp3info.h"

extern char *typegenre [128];


/*
 * Check if Break is given, and if so, clean term and exit, else pad the input.
 */

char *checkinput(int c, char *string, int length) {
        if (c == CTRL_C) { endwin(); exit(0); }
        return string;
}

/*
 * Get the cursor to the right place, and get input
 */

char *curs_addparam(WINDOW *win, int line, int length, char *buf) {
        int c;
        char string[TEXT_FIELD_LEN];

        wmove(win,line,16);
        c = wgetnstr(win,string,length);
        strcpy(buf,checkinput(c,string,length));
        return buf;
}

/*
 * Bring up the curses window, then get and parse input, and build the tag.
 */

void tagedit_curs(char *filename, int filenum, int fileoutof, struct id3tag *tag) {
WINDOW *mainwin;
   char line[50], *genre, genre_text[30];
   unsigned int genre_num;
   line[0]='\0';

   initscr(); cbreak(); noecho();
   nonl();
   mainwin = newwin(8,COLS,5,0);
   intrflush(mainwin, FALSE);
   keypad(mainwin, TRUE);
   CenterText(1,VERSION);
   CenterText(2,"Ricardo Cerqueira <rmc@rccn.net>, 2000");
#ifdef HAVE_TOUCHWIN
   touchwin(mainwin);
#endif
   box(mainwin, 0, 0);
   wmove(mainwin,1,1); 
   wprintw(mainwin,"Song Title:    %s",tag->title);
   wmove(mainwin,2,1); 
   wprintw(mainwin,"Artist Name:   %s",tag->artist);
   wmove(mainwin,3,1); 
   wprintw(mainwin,"Album Name:    %s",tag->album);
   wmove(mainwin,4,1); 
   wprintw(mainwin,"Year:          %s",tag->year);
   wmove(mainwin,5,1); 
   wprintw(mainwin,"Comment:       %s",tag->comment);
   wmove(mainwin,6,1);
   if(tag->genre[0] < 127) {
	genre=typegenre[tag->genre[0]];
   } else {
	genre="";
   }
   strcpy(genre_text,genre);
   wprintw(mainwin,"Genre:         %s",genre);
   wmove(mainwin,7,4);
   wprintw(mainwin," (%d/%d) %s ",filenum,fileoutof,filename);
   wmove(mainwin,7,COLS-22);
   wprintw(mainwin," Press ^C to quit ");
   refresh();
   echo();
   curs_addparam(mainwin,1,30,line);
   strncpy(tag->title,line,strlen(line));
   curs_addparam(mainwin,2,30,line);
   strncpy(tag->artist,line,strlen(line));
   curs_addparam(mainwin,3,30,line);
   strncpy(tag->album,line,strlen(line));
   curs_addparam(mainwin,4,4,line);
   strncpy(tag->year,line,strlen(line));
   curs_addparam(mainwin,5,30,line);
   strncpy(tag->comment,line,strlen(line));
   curs_addparam(mainwin,6,30,line);
   strncpy(genre_text,line,strlen(line) + (strlen(line) < strlen (genre_text) ? 0 : 1));
   endwin();
   if((genre_num=get_genre(genre_text)) < 256) {
	tag->genre[0] = (unsigned char) genre_num;
   } else {
	printf("ERROR - '%s' is not a valid genre name or number.\nUse '-G' for a list of valid genres.\n",genre_text);
   }

}



mp3info.c
/*
 * mp3info.c
 *
 * MP3info's main file. MP3info is an ID3 tag editor.
 *
 * Original Author:
 *       Ricardo Cerqueira      <rmc@rccn.net>
 *
 * This software is covered by the GNU public license, which should be in a file named
 * LICENSE acompanying this.
 *
 */

#include "mp3info.h"
#include <sys/stat.h>

char *typegenre [128] = 
           { 
           "Blues","Classic Rock","Country","Dance","Disco", 
           "Funk","Grunge","Hip-Hop","Jazz","Metal", 
           "New Age","Oldies","Other","Pop","R&B", 
           "Rap","Reggae","Rock","Techno","Industrial", 
           "Alternative","Ska","Death Metal","Pranks","Soundtrack", 
           "Euro-Techno","Ambient","Trip-Hop","Vocal","Jazz+Funk", 
           "Fusion","Trance","Classical","Instrumental","Acid", 
           "House","Game","Sound Clip","Gospel","Noise", 
           "AlternRock","Bass","Soul","Punk","Space", 
           "Meditative","Instrumental Pop","Instrumental Rock","Ethnic","Gothic", 
           "Darkwave","Techno-Industrial","Electronic","Pop-Folk","Eurodance", 
           "Dream","Southern Rock","Comedy","Cult","Gangsta", 
           "Top 40","Christian Rap","Pop/Funk","Jungle","Native American", 
           "Cabaret","New Wave","Psychadelic","Rave","Showtunes", 
           "Trailer","Lo-Fi","Tribal","Acid Punk","Acid Jazz", 
           "Polka","Retro","Musical","Rock & Roll","Hard Rock", 
           "Folk","Folk-Rock","National Folk","Swing","Fast Fusion", 
           "Bebob","Latin","Revival","Celtic","Bluegrass", 
           "Avantgarde","Gothic Rock","Progressive Rock","Psychedelic Rock","Symphonic Rock", 
           "Slow Rock","Big Band","Chorus","Easy Listening","Acoustic", 
           "Humour","Speech","Chanson","Opera","Chamber Music", 
           "Sonata","Symphony","Booty Bass","Primus","Porn Groove", 
           "Satire","Slow Jam","Club","Tango","Samba", 
           "Folklore","Ballad","Power Ballad","Rhythmic Soul","Freestyle", 
           "Duet","Punk Rock","Drum Solo","Acapella","Euro-House", 
           "Dance Hall","Undefined","Undefined"
}; 




int main(int argc, char *argv[]) {
FILE  *fp;
int c, i, interactive = 0, view_only=1,delete_tag=0,file_open,retcode=0,has_tag=0;
int new_genre=0,firstfilearg;
struct id3tag tag,new_tag;
char fullline[129], fbuf[4];
char error_msg[256];
unsigned char sig[2];
unsigned int g;
struct stat filestat;


   new_tag.title[0]=new_tag.artist[0]=new_tag.album[0]=new_tag.year[0]=
   new_tag.comment[0]=new_tag.genre[0]=0;


   if (argc < 2 ) /* Only command is given. Short help */ {
	   printf("%s\r\n  Usage: %s [-h | -G] (-i | -d | -t title -a artist -l album -y year 
	-c comment -g genre) MP3_FILE [ MP3_FILE [ ... ]]\r\n",VERSION,argv[0]); 
	   return 0;
   }


   while ((c=getopt(argc,argv,"vhGidt:a:l:y:c:g:"))!=-1) {

	switch(c) {
		case 'v': /* View mode automatic when no changes made to ID3 tag */
			break;
		case 'h':
			display_help(); return 0;
			break;
		case 'G':
			display_genres(); return 0;
			break;
		case 'i':
			view_only=0;
			interactive = 1;
			break;
		case 'd':
			view_only=0;
			delete_tag=1;
			break;
		case 't':
			strncpy(new_tag.title,optarg,TEXT_FIELD_LEN);
			view_only=0;
			break;
		case 'a':
			strncpy(new_tag.artist,optarg,TEXT_FIELD_LEN);
			view_only=0;
			break;
		case 'l':
			strncpy(new_tag.album,optarg,TEXT_FIELD_LEN);
			view_only=0;
			break;
		case 'y':
			strncpy(new_tag.year,optarg,INT_FIELD_LEN);
			view_only=0;
			break;
		case 'c':
			strncpy(new_tag.comment,optarg,TEXT_FIELD_LEN);
			view_only=0;
			break;
		case 'g':
			g=get_genre(optarg);
			if(g <= 255) {
				new_tag.genre[0] = (unsigned char) g;
				new_genre=1;
				view_only=0;
			} else {
				fprintf(stderr,"Error: '%s' is not a recognized genre name or number.\n",optarg);
				fprintf(stderr,"Use the '-G' option to see a list of valid genre names and numbers\n");
				retcode |= 6;
				return retcode;
			}
			sscanf(optarg,"%u",&g);
			break;
	}
   }

   firstfilearg=optind;

   for(i=optind;i < argc; i++) { /* Iterate over all filenames */
      file_open=0;
      if (view_only == 1) { 
        if ( !( fp=fopen(argv[i],"r") ) ) {
  	        sprintf(error_msg,"Error opening MP3: %s",argv[i]);
                perror(error_msg);
		retcode |= 1;
        } else {
		file_open=1;
	}
      } else {
        if ( !( fp=fopen(argv[i],"rb+") ) ) {
  	        sprintf(error_msg,"Error opening MP3: %s",argv[i]);
                perror(error_msg);
	        retcode |= 1;
        } else {
		file_open=1;
	}
      }

      if(file_open == 1) {

         /* Let's see if it's an MP3. SMJ 06/1999 */  
 
         fread(sig,sizeof(sig),1,fp);                         
         sig[0] &= 0xff;
         sig[1] &= 0xf0;                                                         
         if(!((sig[0] == 0xff) && (sig[1] == 0xf0))) {
                 fprintf(stderr,"%s is not an MP3 file!\n",argv[i]);        
                 retcode |= 2;
         } else { /* Yup, it's an MP3 */

            /* Go to EOF and check if there's already a tag */

            if ( fseek(fp, -128, SEEK_END )) { 
		retcode |= 4;
	    } else {
               fread(fbuf,1,3,fp); fbuf[3] = '\0';
   	       tag.title[0]=tag.artist[0]=tag.album[0]=tag.year[0]=tag.comment[0]=0;
	       tag.genre[0]=255;
               if (!strcmp("TAG",fbuf)) {
            	   /* Holy cow, Batman! I found something! Better save it for later use... */
		   has_tag=1;
            	   fseek(fp, -125, SEEK_END); 

            	   fread(tag.title,1,30,fp); tag.title[30] = '\0';
            	   fread(tag.artist,1,30,fp); tag.artist[30] = '\0';
            	   fread(tag.album,1,30,fp); tag.album[30] = '\0';
            	   fread(tag.year,1,4,fp); tag.year[4] = '\0';
            	   fread(tag.comment,1,30,fp); tag.comment[30] = '\0';
            	   fread(tag.genre,1,1,fp);
            	   /* freads moved the file pointer. Get it back to the beginning of the tag so it 	*
            	    * will be overwritten 								*/
                   fseek(fp, -128, SEEK_END);
               } else {
            	   /* No tags here. Back to the EOF to append the new one */
		   has_tag=0;
            	   fseek(fp, 0, SEEK_END); 
               }

	       if(view_only) {
	          if(has_tag) {
		     view_tag(argv[i],&tag);
                  } else {
   	       	     printf("No ID3 found in MP3 file %s\n",argv[i]);
                  }
   
	       } else {

	          if(new_tag.title[0]) {
			strncpy(tag.title,new_tag.title,TEXT_FIELD_LEN);
	          }

	          if(new_tag.artist[0]) {
			strncpy(tag.artist,new_tag.artist,TEXT_FIELD_LEN);
	          }

	          if(new_tag.album[0]) {
			strncpy(tag.album,new_tag.album,TEXT_FIELD_LEN);
	          }

	          if(new_tag.comment[0]) {
			strncpy(tag.comment,new_tag.comment,TEXT_FIELD_LEN);
	          }

	          if(new_tag.year[0]) {
			strncpy(tag.year,new_tag.year,INT_FIELD_LEN);
	          }

	          if(new_genre) {
			tag.genre[0]=new_tag.genre[0];
	          }
  		  
		  if( interactive ) {
	 	     tagedit_curs(argv[i],i-firstfilearg+1,argc-firstfilearg,&tag);
		  }
                  
                  buildtag(fullline,&tag);
               
                  /* Finally! Get it done! */
                  if(!delete_tag) {
			fwrite(fullline,1,128,fp);
                  }

	       }
            }

	 }                                                    

         fclose(fp);
         if(delete_tag && has_tag) {
		stat(argv[i],&filestat);
		truncate(argv[i],filestat.st_size-128);
	 }
      }



   }

   if(optind == argc) {
	fprintf(stderr,"No MP3 files specified!\n");
	retcode |= 8;
   }

   return retcode;

}





textfunc.h
/*
 * textfunc.h
 *
 * Original Author:
 *       Ricardo Cerqueira      <rmc@rccn.net>
 *
 * 
 * This software is covered by the GNU public license, which should be in a file named
 * LICENSE acompanying this.
 *
 */

#define CTRL_C		'\003'
#define TEXT_FIELD_LEN	30
#define INT_FIELD_LEN	4

struct id3tag {
	char title[31];
	char artist[31];
	char album[31];
	char year[5];
	char comment[31];
	unsigned char genre[1];
};

void CenterText(int line, char *text);
char *checkinput(int c, char *string, int length);
void pad(char *string, int length);
char *curs_addparam(WINDOW *win, int line, int length, char *buf);
void buildtag(char *buf, struct id3tag *tag);
void display_help();
void view_tag(char *filename,struct id3tag *tag);
void display_genres(void);
unsigned int get_genre (char *genre);




gmp3info.c
/*
 * gmp3info.c
 *
 * GTK MP3info's main file. MP3info is an ID3 tag editor.
 *
 * Original Author:
 *       Ricardo Cerqueira      <rmc@rccn.net>
 *
 * This software is covered by the GNU public license, which should be in a file named
 * LICENSE acompanying this.
 *
 */

#include <gtk/gtk.h>
#include "mp3info.h"

gchar *gtext_genre(unsigned char genre);
unsigned char gget_genre (char *genre);

char *typegenre[129] = 
           { 
           "Blues","Classic Rock","Country","Dance","Disco", 
           "Funk","Grunge","Hip-Hop","Jazz","Metal", 
           "New Age","Oldies","Other","Pop","R&B", 
           "Rap","Reggae","Rock","Techno","Industrial", 
           "Alternative","Ska","Death Metal","Pranks","Soundtrack", 
           "Euro-Techno","Ambient","Trip-Hop","Vocal","Jazz+Funk", 
           "Fusion","Trance","Classical","Instrumental","Acid", 
           "House","Game","Sound Clip","Gospel","Noise", 
           "AlternRock","Bass","Soul","Punk","Space", 
           "Meditative","Instrumental Pop","Instrumental Rock","Ethnic","Gothic", 
           "Darkwave","Techno-Industrial","Electronic","Pop-Folk","Eurodance", 
           "Dream","Southern Rock","Comedy","Cult","Gangsta", 
           "Top 40","Christian Rap","Pop/Funk","Jungle","Native American", 
           "Cabaret","New Wave","Psychadelic","Rave","Showtunes", 
           "Trailer","Lo-Fi","Tribal","Acid Punk","Acid Jazz", 
           "Polka","Retro","Musical","Rock & Roll","Hard Rock", 
           "Folk","Folk-Rock","National Folk","Swing","Fast Fusion", 
           "Bebob","Latin","Revival","Celtic","Bluegrass", 
           "Avantgarde","Gothic Rock","Progressive Rock","Psychedelic Rock","Symphonic Rock", 
           "Slow Rock","Big Band","Chorus","Easy Listening","Acoustic", 
           "Humour","Speech","Chanson","Opera","Chamber Music", 
           "Sonata","Symphony","Booty Bass","Primus","Porn Groove", 
           "Satire","Slow Jam","Club","Tango","Samba", 
           "Folklore","Ballad","Power Ballad","Rhythmic Soul","Freestyle", 
           "Duet","Punk Rock","Drum Solo","Acapella","Euro-House", 
           "Dance Hall","Undefined","Undefined",""
}; 

int alphagenreindex[] = {
128,123,74,73,34,99,40,20,26,90,116,41,85,
96,89,0,107,65,88,104,102,97,61,1,32,
112,57,2,58,125,3,50,22,4,55,122,120,
98,52,48,124,25,54,84,81,115,80,119,5,
30,36,59,38,91,49,6,79,7,35,100,19,
46,47,33,29,8,63,86,71,45,9,77,82,
64,10,66,39,11,103,12,75,53,62,13,109,
117,23,108,92,67,93,121,43,14,15,68,16,
76,87,118,78,17,114,110,69,21,111,95,105,
42,37,24,56,44,101,83,94,106,113,51,18,
60,70,31,72,27,126,28
};

GtkWidget	*id3win,*id3win_frame_global;
GtkWidget	*id3win_frame_title,*id3win_frame_artist,*id3win_frame_album;
GtkWidget	*id3win_frame_year,*id3win_frame_comment, *id3win_frame_albyear;
GtkWidget	*id3win_frame_genre;
GtkWidget	*id3win_text_title,*id3win_text_artist,*id3win_text_album;
GtkWidget	*id3win_text_year,*id3win_text_comment,*id3win_combo_genre;
GtkWidget	*id3win_frame_buttons,*id3win_ok_button,*id3win_cancel_button;
GList 		*genrelist=NULL;
gchar		title[31]="",artist[31]="",album[31]="",year[4]="",comment[31]="",fbuf[4];

char fulltag[128];
unsigned char genre;
FILE  *fp;
unsigned char sig[2];

char *gpad (char *string, int length) {
	while(strlen(string)<length) {
		strcat(string," ");
	}
	return string;
}

void unpad (char *string) {
	int x;
	for (x=(strlen(string)-1); x>=0; x--) {
		if (string[x]!=' ') {
			return;
		} else {
			string[x]='\0';
		}
	}
}

void exit_save( GtkWidget *widget, GtkWidget *data) {
   	char tmp[31];
   	strcpy(fulltag,"TAG");
   	strcpy(tmp,gtk_entry_get_text(GTK_ENTRY(id3win_text_title)));
   	strcat(fulltag,gpad(tmp,30));
   	strcpy(tmp,gtk_entry_get_text(GTK_ENTRY(id3win_text_artist)));
   	strcat(fulltag,gpad(tmp,30));
   	strcpy(tmp,gtk_entry_get_text(GTK_ENTRY(id3win_text_album)));
   	strcat(fulltag,gpad(tmp,30));
   	strcpy(tmp,gtk_entry_get_text(GTK_ENTRY(id3win_text_year)));
   	strcat(fulltag,gpad(tmp,4));
   	strcpy(tmp,gtk_entry_get_text(GTK_ENTRY(id3win_text_comment)));
   	strcat(fulltag,gpad(tmp,30));
        strcpy(tmp,gtk_entry_get_text(GTK_ENTRY(GTK_COMBO(id3win_combo_genre)->entry)));
        fulltag[127]=gget_genre(tmp);
   	fwrite(fulltag,1,128,fp);
   	fclose(fp);
   	gtk_main_quit();
}

int main(int argc, char *argv[] ) {
    
   int i;

   if(argc==2) { 
    if ( !( fp=fopen(argv[1],"rb+") ) ) {
            perror("Error opening MP3");
            return 0;
    }

    fread(sig,sizeof(sig),1,fp);
    sig[0] &= 0xff;                                                          
    sig[1] &= 0xf0;                                                          
    if(!((sig[0] == 0xff) && (sig[1] == 0xf0))) {  
            perror("File is not an MP3!");        
            return 0;
    }

    if ( fseek(fp, -128, SEEK_END )) { return 0; }
    fread(fbuf,1,3,fp); fbuf[3] = '\0';
    if (!strcmp("TAG",fbuf)) {
 	   fseek(fp, -125, SEEK_END); 
 	   fread(title,1,30,fp); unpad(title);
 	   fread(artist,1,30,fp); unpad(artist);
 	   fread(album,1,30,fp); unpad(album);
 	   fread(year,1,4,fp); unpad(year);
 	   fread(comment,1,30,fp); unpad(comment);
	   fread(&genre,1,1,fp);
           fseek(fp, -128, SEEK_END);
    } else {
 	   fseek(fp, 0, SEEK_END); 
    }
  
	gtk_init (&argc, &argv);

	id3win=gtk_window_new(GTK_WINDOW_DIALOG);
	gtk_window_set_title(GTK_WINDOW(id3win),"MP3info - An ID3 tag editor");
	gtk_window_set_policy (GTK_WINDOW(id3win), FALSE,FALSE,FALSE);
	gtk_widget_set_usize(id3win,450,340);
	gtk_container_border_width(GTK_CONTAINER(id3win),10);
	gtk_signal_connect(GTK_OBJECT(id3win), "delete_event", (GtkSignalFunc) gtk_exit, NULL);
	id3win_frame_global=gtk_vbox_new(FALSE,10);
	
        id3win_frame_title=gtk_frame_new("Title");
        gtk_container_border_width(GTK_CONTAINER(id3win_frame_title),5);
        id3win_text_title=gtk_entry_new_with_max_length(30);
	gtk_entry_set_text(GTK_ENTRY(id3win_text_title),title);
        gtk_container_add(GTK_CONTAINER(id3win_frame_title),id3win_text_title);
	gtk_box_pack_start(GTK_BOX(id3win_frame_global),id3win_frame_title,TRUE,TRUE,0);

        id3win_frame_artist=gtk_frame_new("Artist");
        gtk_container_border_width(GTK_CONTAINER(id3win_frame_artist),5);
        id3win_text_artist=gtk_entry_new_with_max_length(30);
	gtk_entry_set_text(GTK_ENTRY(id3win_text_artist),artist);
        gtk_container_add(GTK_CONTAINER(id3win_frame_artist),id3win_text_artist);
	gtk_box_pack_start(GTK_BOX(id3win_frame_global),id3win_frame_artist,TRUE,TRUE,0);

	id3win_frame_albyear=gtk_hbox_new(FALSE,30);
        id3win_frame_album=gtk_frame_new("Album");
        gtk_container_border_width(GTK_CONTAINER(id3win_frame_album),5);
        id3win_text_album=gtk_entry_new_with_max_length(30);
	gtk_entry_set_text(GTK_ENTRY(id3win_text_album),album);
        gtk_container_add(GTK_CONTAINER(id3win_frame_album),id3win_text_album);
	gtk_box_pack_start(GTK_BOX(id3win_frame_albyear),id3win_frame_album,TRUE,TRUE,0);

        id3win_frame_year=gtk_frame_new("Year");
	gtk_widget_set_usize(id3win_frame_year,2,0);
        gtk_container_border_width(GTK_CONTAINER(id3win_frame_year),5);
        id3win_text_year=gtk_entry_new_with_max_length(4);
	gtk_entry_set_text(GTK_ENTRY(id3win_text_year),year);
        gtk_container_add(GTK_CONTAINER(id3win_frame_year),id3win_text_year);
	gtk_box_pack_start(GTK_BOX(id3win_frame_albyear),id3win_frame_year,TRUE,TRUE,0);
	gtk_box_pack_start(GTK_BOX(id3win_frame_global),id3win_frame_albyear,TRUE,TRUE,0);

        id3win_frame_comment=gtk_frame_new("Comment");
        gtk_container_border_width(GTK_CONTAINER(id3win_frame_comment),5);
        id3win_text_comment=gtk_entry_new_with_max_length(30);
	gtk_entry_set_text(GTK_ENTRY(id3win_text_comment),comment);
        gtk_container_add(GTK_CONTAINER(id3win_frame_comment),id3win_text_comment);
	gtk_box_pack_start(GTK_BOX(id3win_frame_global),id3win_frame_comment,TRUE,TRUE,0);

        id3win_frame_genre=gtk_frame_new("Genre");
        gtk_container_border_width(GTK_CONTAINER(id3win_frame_genre),5);
        id3win_combo_genre=gtk_combo_new();
        for(i=0;i<128;i++) {
		genrelist = g_list_append(genrelist, typegenre[alphagenreindex[i]]);
	}
        gtk_combo_set_popdown_strings(GTK_COMBO(id3win_combo_genre),genrelist);
        gtk_entry_set_text(GTK_ENTRY(GTK_COMBO(id3win_combo_genre)->entry), gtext_genre(genre));
        gtk_container_add(GTK_CONTAINER(id3win_frame_genre),id3win_combo_genre);
	gtk_box_pack_start(GTK_BOX(id3win_frame_global),id3win_frame_genre,TRUE,TRUE,0);

	id3win_frame_buttons=gtk_hbox_new(TRUE,30);
	id3win_ok_button=gtk_button_new_with_label("OK");
        gtk_box_pack_start(GTK_BOX(id3win_frame_buttons),id3win_ok_button,TRUE,TRUE,0);
	id3win_cancel_button=gtk_button_new_with_label("Cancel");
        gtk_box_pack_start(GTK_BOX(id3win_frame_buttons),id3win_cancel_button,TRUE,TRUE,0);
	gtk_widget_set_usize(id3win_frame_buttons,30,20);
	gtk_box_pack_start(GTK_BOX(id3win_frame_global),id3win_frame_buttons,TRUE,TRUE,0);

        gtk_container_add(GTK_CONTAINER(id3win),id3win_frame_global);

	gtk_signal_connect (GTK_OBJECT (id3win_ok_button), "clicked", GTK_SIGNAL_FUNC (exit_save), NULL);
	gtk_signal_connect (GTK_OBJECT (id3win_cancel_button), "clicked", GTK_SIGNAL_FUNC (gtk_exit), NULL);
	
	gtk_widget_show_all(id3win);
	gtk_main();

	return(0);
   } else {
	   g_print("You must supply an MP3 file as an argument!\n");
	   return(0);
   }

			
}

gchar *gtext_genre(unsigned char genre) {
   int genre_num = (int) genre;

   if(genre_num < 127) {
	return((gchar *)typegenre[genre_num]);
   } else {
	return((gchar *)"");
   }
}

unsigned char gget_genre (char *genre) {
	int num_genre=0;

        if(genre[0] == '\0') { return 255; }

	sscanf(genre,"%u",&num_genre);
	if(num_genre == 0) {
		if(genre[0] != '0') {
			while(num_genre++ < 128) {
				if(!strcasecmp(genre,typegenre[num_genre-1])) {
					return num_genre-1;
				}
			}
			num_genre=256;
		}		
	}

	if(num_genre < 0 || num_genre > 126) {
		num_genre=255;
	}
	return (unsigned char) num_genre;
}


